egs/wsj/s5/steps/nnet3/multilingual/allocate_multilingual_examples.py

```diff
@@ -0,0 +1,264 @@
+
+          lang_len[lang_id] = lang_len[lang_id] - args.minibatch_size
+          num_egs = num_egs + args.minibatch_size;
+          # If the num of remaining egs in each lang is less than minibatch_size,
+          # they are discarded.
+          if lang_len[lang_id] < args.minibatch_size:
+            lang_len[lang_id] = 0
+            print("Run out of data for language {0}".format(lang_id))
+        else:
+          print("Run out of data for all languages.")
+          break
+      all_egs.append(this_egs)
+      archfile.close()
+
+  # combine examples across all jobs correspond to each archive.
+  for archive in range(num_archives):
+    print("Processing archive {0} by combining all jobs.".format(archive + 1))
+    this_ranges = []
+    f = open(args.egs_dir + "/temp/" + args.prefix + "ranges." + str(archive + 1), "w")
+    o = open(args.egs_dir + "/" + args.prefix + "output." + str(archive + 1), "w")
+    w = open(args.egs_dir + "/" + args.prefix + "weight." + str(archive + 1), "w")
+    scp_per_archive_file = open(args.egs_dir + "/" + args.prefix + "egs." + str(archive + 1), "w")
+
+    # check files befor writing.
+    if f is None:
+      raise Exception("Error opening file " + args.egs_dir + "/temp/" + args.prefix + "ranges." + str(job + 1))
+    if o is None:
+      raise Exception("Error opening file " + args.egs_dir + "/" + args.prefix + "output." + str(job + 1))
+    if w is None:
+      raise Exception("Error opening file " + args.egs_dir + "/" + args.prefix + "weight." + str(job + 1))
+    if scp_per_archive_file is None:
+      raise Exception("Error opening file " + args.egs_dir + "/" + args.prefix + "egs." + str(archive + 1), "w")
+
+    for job in range(args.num_jobs):
+      # combine egs.job.archive.scp across all jobs.
+      scp = args.egs_dir + "/temp/" + args.prefix + "scp." + str(job + 1) + "." + str(archive + 1)
+      with open(scp,"r") as scpfile:
+        for line in scpfile:
+          scp_line = line.splitlines()[0].split()
+          print("{0} {1}".format(scp_line[0], scp_line[1]), file=scp_per_archive_file)
+          print("{0} output-{1}".format(scp_line[0], scp_line[2]), file=o)
+          print("{0} {1}".format(scp_line[0], lang2weight[int(scp_line[2])]), file=w)
+      os.remove(scp)
+
+      # combine ranges.* across all jobs for archive
+      for (lang_id, start_eg_line, num_egs) in all_egs[num_archives * job + archive]:
+        this_ranges.append((lang_id, start_eg_line, num_egs))
+
+    # write ranges.archive
+    for (lang_id, start_eg_line, num_egs) in this_ranges:
+      print("{0} {1} {2}".format(lang_id, start_eg_line, num_egs), file=f)
+
+    scp_per_archive_file.close()
+    f.close()
+    o.close()
+    w.close()
+  print("allocate_multilingual_examples.py finished generating " + args.prefix + "egs.*.scp and " + args.prefix +  "ranges.* and " + args.prefix + "output.*" + args.prefix + "weight.* files")
```

Comment:
By Google standards, you should use format instead of using + to concatenate string.